### PR TITLE
add uftrace-test.md

### PR DESCRIPTION
**Pull request by lee5black**
[https://github.com/namhyung/uftrace/pull/604](url)

I will send you the revised version `uftrace-test.md`!

### CODE REVIEW

**Comment by namhyung** [comment]
The link is broken and you can only write the issue number preceded by a sharp (hash) like `#604`.  Also you don't need to close/create issue for a new version.  Please just (force) push to the same branch and leave a comment in the PR.

File: doc/uftrace-test.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+`````
+root@ubuntu:~/uftrace/tests$ ./runtest.py 001 -p -v -O1
+Test case                 pg
+------------------------: O1
+build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -pg -O1  s-abc.c   
+test command: ../uftrace --no-pager --no-event -L.. t-abc
+
+`````
+
+## ADD TEST CASE
```
Comment (namhyung) [inline doc/uftrace-test.md]: What about "## How to add test case" or "## WRITING TEST CASE"?

File: doc/uftrace-test.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+	return 0;
+}
+
+void TEST(void){
+	printf("USER_TEST_CODE\n");
+}
+```
+
+**second > Apply the default gcc and uftrace options in runtest.py to get the following results:**
+(If you want, you can optionally modify your runtest.py to give your own default gcc option or uftrace.)
```
Comment (namhyung) [inline doc/uftrace-test.md]: It'd be nice to show an example of that.

File: doc/uftrace-test.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+            [  6053] | main() {
+            [  6053] |   TEST() {
+ 131.151 us [  6053] |     printf();
+ 132.454 us [  6053] |   } /* TEST */
+ 132.658 us [  6053] | } /* main */
+ """)
+
+        def runcmd(self):
+            return '%s -N c %s' % (TestBase.uftrace_cmd, 't-USER_TEST')
+```
```
Comment (namhyung) [inline doc/uftrace-test.md]: At the minimum, you don't even need the `runcmd()` function.  The runtest.py will run uftrace based on your test name (given on `__init__()`) with no option.

**Comment by lee5black** [comment]
Thank you for your feedback. This time, I will not send a new PR, but I will send a PR through `force` again.

**Comment by honggyukim** [comment]
@lee5black Thank you very much for your work. It's worth to have a document for testing.  But I think it doesn't have to follow other man page document format and it seems a bit verbose. I think it's better to have shorter sentences and examples.

I don't know if @lee5black has a permission to write the doc in wiki, but if we once have this document in wiki page, we can freely modify it together and it will make the document better and clearer.

**Comment by lee5black** [comment]
@honggyukim I checked my permission to write a document with wiki, but I think it is not. How shall I do it?

**Comment by namhyung** [comment]
@lee5black I don't want to change the wiki permission due to possible spams.  I think we can discuss this in the issue page and move it to wiki once it's finished.

File: doc/uftrace-test.md [hunk 4]
```diff
@@ -39,12 +39,12 @@ This program supports multiple tests and can be found through the options below.
 
 -t TIMEOUT, --timeout TIMEOUT : fail test if it runs more than TIMEOUT seconds
 
-## EXAMPEL
+## EXAMPLE
```
Comment (lee5black) [inline doc/uftrace-test.md]: done.

**Comment by honggyukim** [comment]
You don't have to say sorry for that but please combine those 4 commits into a single commit, then force push it again.

**Comment by lee5black** [comment]
Okay. Thanks.

**Comment by lee5black** [comment]
I sent it. Please check.

**Comment by honggyukim** [comment]
Hmm.. you should read something like how to write a good commit message.
- For example, https://meetup.toast.com/posts/106 (written in Korean)

Please remove empty whitespaces at the end of line. If you run `git diff --check HEAD^`, then you will see something like below:
```
$ git diff --check HEAD^
doc/uftrace-test.md:19: trailing whitespace.
+[case]  : If nothing is entered, the test progresses for the entire test case.
doc/uftrace-test.md:25: trailing whitespace.
+-f FLAGS, --profile-flags FLAGS : comma separated list of compiler profiling flags.
doc/uftrace-test.md:33: trailing whitespace.
+-d, --diff  : show diff result if not matched.
doc/uftrace-test.md:72: trailing whitespace.
+build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -pg -O1  s-abc.c
doc/uftrace-test.md:108: trailing whitespace.
+build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -finstrument-functions -O1  s-abc.c
doc/uftrace-test.md:160: trailing whitespace.
+program instrumentation option reference :
doc/uftrace-test.md:167: trailing whitespace.
+build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -fprofile-arcs -O1  s-abc.c
doc/uftrace-test.md:176: trailing whitespace.
+build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -pg -O1  s-abc.c
doc/uftrace-test.md:184: trailing whitespace.
+build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -finstrument-functions -O1  s-abc.c
doc/uftrace-test.md:192: trailing whitespace.
+build command: gcc -o t-abc -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -pg -O1  s-abc.c
doc/uftrace-test.md:206: trailing whitespace.
+$ cat s-USER_TEST.c
doc/uftrace-test.md:268: trailing whitespace.
+build command: gcc -o t-USER_TEST -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -pg -O1  s-USER_TEST.c
doc/uftrace-test.md:293: trailing whitespace.
+build command: gcc -o t-USER_TEST -fno-inline -fno-builtin -fno-ipa-cp -fno-omit-frame-pointer -D_FORTIFY_SOURCE=0  -finstrument-functions -O1  s-USER_TEST.c
```
We will think what to do about wiki permission as well.

**Comment by honggyukim** [comment]
It also contains some broken characters, which look like `<U+200B>`.
```diff
+## OPTIONS
+
+[case]  : If nothing is entered, the test progresses for the entire test case.
+
+<U+200B>              Test number, test case name (partial): Test progress for the test case
+
+-h, --help  : show this help message and exit
+
+-f FLAGS, --profile-flags FLAGS : comma separated list of compiler profiling flags.
+
+-O OPTS, --optimize-levels OPTS : compiler optimization levels
+
+-p, --profile-pg : profiling with -pg option
+
+-i, --instrument-functions : profiling with -finstrument-functions option
+
+-d, --diff  : show diff result if not matched.
+<U+200B>                  The comparison result of uftrace execution is stored in ` t[testnumber]_testname.py`
+
+-v, --verbose : show internal command and result for debugging (Up to 3 additional)
```

**Comment by namhyung** [comment]
Changed the wiki permission.

**Comment by lee5black** [comment]
Thank you. Written briefly on the wiki!

**Comment by lee5black** [comment]
@honggyukim  

default

I delete trailing whitespace. and  modified my commit message.

image

Can I make more changes?

**Comment by honggyukim** [comment]
> Thank you. Written briefly on the wiki!

Thanks a lot! We will gradually improve the document in the wiki page.

> Can I make more changes?

I just made the comment to let you know how you can fix this kind of commit next time. As it's already written in the wiki page, this PR can be closed. Thanks you.

**Comment by lee5black** [comment]
Thank you ! `uftrace test.md` are not merged and they replaced by wikis? If it is, I will quit PR now!

**Pull request closed without merging**